ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PoWSolver.java

```diff
@@ -149,22 +155,50 @@ public void cancel() {
 
   public boolean submitSolution(final PoWSolution solution) {
     final Optional<PoWSolverJob> jobSnapshot = currentJob;
+    PoWSolverJob jobToTestWith = null;
     if (jobSnapshot.isEmpty()) {
       LOG.debug("No current job, rejecting miner work");
       return false;
     }
 
-    final PoWSolverJob job = jobSnapshot.get();
-    final PoWSolverInputs inputs = job.getInputs();
-    if (!inputs.getPrePowHash().equals(solution.getPowHash())) {
-      LOG.debug("Miner's solution does not match current job");
+    PoWSolverJob headJob = jobSnapshot.get();
+    if (headJob.getInputs().getPrePowHash().equals(solution.getPowHash())) {
+      LOG.debug("Head job matches the solution pow hash {}", solution.getPowHash());
+      jobToTestWith = headJob;
+    }
+    if (jobToTestWith == null) {
+      PoWSolverJob ommerCandidate = currentJobs.get(solution.getPowHash());
+      if (ommerCandidate != null) {
+        long distanceToHead =
+            headJob.getInputs().getBlockNumber() - ommerCandidate.getInputs().getBlockNumber();
+        LOG.debug(
+            "Found ommer candidate {} with block number {}, distance to head {}",
+            solution.getPowHash(),
+            ommerCandidate.getInputs().getBlockNumber(),
+            distanceToHead);
+        if (distanceToHead <= MAX_OMMER_DEPTH) {
+          jobToTestWith = ommerCandidate;
+        } else {
+          LOG.debug("Discarded ommer solution as too far from head {}", distanceToHead);
+        }
+      }
+    }
+    if (jobToTestWith == null) {
+      LOG.debug("No matching job found for hash {}, rejecting solution", solution.getPowHash());
       return false;
     }
+    if (jobToTestWith.isDone()) {
+      LOG.debug("Matching job found for hash {}, but already solved", solution.getPowHash());
+      return false;
+    }
+    final PoWSolverInputs inputs = jobToTestWith.getInputs();
+
     final Optional<PoWSolution> calculatedSolution = testNonce(inputs, solution.getNonce());
 
     if (calculatedSolution.isPresent()) {
       LOG.debug("Accepting a solution from a miner");
-      currentJob.get().solvedWith(calculatedSolution.get());
+      currentJobs.remove(solution.getPowHash());
```

Comment:
Does the miner want multiple ommers minable?  Removing on the first solution precludes that.

Reply:
Right now, we only accept one solution per block (the `solveFor` method blocks until the future completes). This repeats the same mechanism but with ommer blocks. We could potentially choose to allow more than one solution per block, turn those into ommers and replace the proposed block with the block with the best nonce. That's why I wanted a review, I'm just not sure how far to push this.

Reply:
To be clear, that'd be best addressed as a separate PR, because changing `solveFor` feels like a change worth doing on its own, whether we accept mining ommers or not.